### PR TITLE
Add String::Builder#clear

### PR DESCRIPTION
**Pull request by r00ster91**
One of my instance variables in my class is a `String::Builder` and I would really like to clear it efficiently instead of having to build a new `String::Builder` every time I want to build a new string with it.
So I added this which simply sets the bytesize of the builder to zero so the old bytes from the other string build are simply overwritten by the new string build.

```crystal
io = String::Builder.new

Benchmark.ips do |bm|
  bm.report("new") do
    io = String::Builder.new
  end

  bm.report("clear") do
    io.clear
  end
end
```
```
  new   6.86M (145.79ns) (±12.55%)  176 B/op  88.42× slower
clear  606.5M (  1.65ns) (± 6.06%)    0 B/op        fastest
```
This also adds a NOTE to `String.build` which suggests that when you use `String.build` in a loop, that you should use the `String::Builder` class and simply clear the old contents of the builder and then build a new string with the same builder. That's of course much faster than creating a new `String::Builder` everytime which is done by `String.build` in a loop. I did these improvements in `src/mime/media_type.cr`, for example.
```crystal
io = String::Builder.new

Benchmark.ips do |bm|
  bm.report("build") do
    10.times do
      String.build do |io|
        io << "abc"
      end
    end
  end

  bm.report("clear") do
    10.times do
      io << "abc"
      io.to_s
      io.clear
    end
  end
end
```
```
build 516.91k (  1.93µs) (±21.86%)  2083 B/op  19.69× slower
clear  10.18M ( 98.27ns) (±11.60%)     0 B/op        fastest
```

### CODE REVIEW

File: spec/std/string_builder_spec.cr [hunk 1]
```diff
@@ -9,6 +9,17 @@ describe String::Builder do
     str.should eq("123456")
   end
 
+  it "clears" do
+    builder = String::Builder.new
```
Comment (Sija) [inline spec/std/string_builder_spec.cr]:
```suggestion
    io = String::Builder.new
```

Comment (r00ster91) [reply spec/std/string_builder_spec.cr]: Oh whoops, thanks.

**Comment by ysbaddaden** [comment]
AFAIK this is **unsafe** and there is good reason for the builder to not be reusable: the allocated buffer is merely casted to a String. This means that this clear method is modifying the previously generated String (unexpected) and possibly reallocating the buffer (segfault).

**Comment by Sija** [comment]
@ysbaddaden Good call! Shouldn't be possible to realloc buffer with `@buffer = GC.malloc_atomic(capacity.to_u32).as(UInt8*)` call within `#clear` to avoid case you've mentioned?

**Comment by r00ster91** [comment]
Oh I see what you mean:
```crystal
class String::Builder < IO
  def clear
    @bytesize = 0
    @finished = false
  end
end

io = String::Builder.new

io << "hello"

string = io.to_s

io.clear

io << "123"

string # => "123lo"
```
But yes it seems to be possible to fix by reallocating the buffer:
```crystal
class String::Builder < IO
  def clear
    @bytesize = 0
    @finished = false
    @buffer = GC.malloc_atomic(@capacity.to_u32).as(UInt8*)
  end
end

io = String::Builder.new

io << "hello"

string = io.to_s

io.clear

io << "123"

string # => "hello"
```

**Comment by asterite** [comment]
String::Builder shouldn't be public API nor used directly. Use IO::Memory. We should probably add mutating methods to it to make it easy to build and modify a string.

**Comment by ysbaddaden** [comment]
Closing. String::Builder is internal detail to implement `String.build`. Messing with it is dangerous.

As stated by Asterite, please use an IO::Memory object.

**Comment by Sija** [comment]
Anyhow, it seems there's a performance improvement to be done in `MIME::MediaType.parse_impl`.

/cc @straight-shoota

**Comment by straight-shoota** [comment]
Yeah, although it's probably not a huge improvement, in practice continuation parameters are actually quite rare I think. However, there are probably other places in the stdlib as well where we could reuse a string buffer.

**Pull request closed without merging**